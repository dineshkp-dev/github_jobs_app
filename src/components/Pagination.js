import React, { useState, useEffect, useMemo } from 'react';
import { JOBS_PER_PAGE } from '../common/constants';
import '../styles/pagination.css';

function Pagination(props) {
  const { total, currentPage, onPageChange } = props;
  const [totalPages, setTotalPages] = useState(0);

  useEffect(() => {
    if (total && JOBS_PER_PAGE) {
      setTotalPages(Math.ceil(total / JOBS_PER_PAGE));
    }
  }, [total]);

  const paginationItems = useMemo(() => {
    const pages = [];
    for (let i = 1; i <= totalPages; i++) {
      pages.push(
        <div key={i} className='pagination-class-indicator'>
          {currentPage === i ? (
            <div className='pagination-page active-page'>{i}</div>
          ) : (
            <div className='pagination-page'>{i}</div>
          )}
        </div>
      );
    }
    return pages;
  }, [totalPages, currentPage]);

  if (!!!totalPages) {
    return null;
  }

  const isNextButtonActive = () => {
    return !!!(currentPage + 1 > totalPages);
  };

  const isPreviousButtonActive = () => {
    return !!!(currentPage - 1 === 0);
  };

  return (
    <div className='pagination-container'>
      {isPreviousButtonActive() ? (
        <div className='button' onClick={() => onPageChange(currentPage - 1)}>
          <div className='button-text'>Previous</div>
        </div>
      ) : (
        <div className='button-disabled'>
          <div className='button-text'>Previous</div>
        </div>
      )}

      {paginationItems}
      {isNextButtonActive() ? (
        <div className='button' onClick={() => onPageChange(currentPage + 1)}>
          <div className='button-text'>Next</div>
        </div>
      ) : (
        <div className='button-disabled'>
          <div className='button-text'>Next</div>
        </div>
      )}
    </div>
  );
}

export default Pagination;
