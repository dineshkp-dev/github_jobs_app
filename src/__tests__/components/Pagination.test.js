import Pagination from '../../components/Pagination';
import { render, screen, fireEvent, waitFor } from '@testing-library/react';
import '@testing-library/jest-dom/extend-expect';

describe('Pagination snapshot test', () => {
  const totalJobs = 5;
  const currentPage = 1;
  test('renders', () => {
    const { container } = render(
      <Pagination total={totalJobs} currentPage={currentPage} />
    );
    expect(container).toMatchSnapshot();
  });
});

describe('Pagination render test', () => {
  test('renders Home with one page only', () => {
    const totalJobs = 5;
    const currentPage = 2;
    render(<Pagination total={totalJobs} currentPage={currentPage} />);
    expect(screen.getByText('Previous')).toBeInTheDocument();
    expect(screen.getByText('Next')).toBeInTheDocument();
    expect(screen.getByText('1')).toBeInTheDocument();
  });

  test('renders Pagination with multiple pages', () => {
    const totalJobs = 35;
    const currentPage = 2;
    render(<Pagination total={totalJobs} currentPage={currentPage} />);
    expect(screen.getByText('Previous')).toBeInTheDocument();
    expect(screen.getByText('Next')).toBeInTheDocument();
    expect(screen.getByText('1')).toBeInTheDocument();
    expect(screen.getByText('2')).toBeInTheDocument();
  });

  test('renders Pagination with 0 pages', () => {
    const totalJobs = 0;
    const currentPage = 2;
    render(<Pagination total={totalJobs} currentPage={currentPage} />);
  });
});

// fireEvent.click(screen.getByText('Load Greeting'))
describe('Pagination simulate navigation', () => {
  test('navigate next', async () => {
    const totalJobs = 30;
    const currentPage = 1;
    let nextPage = 1;
    const onPageChange = (newPageNumber) => {
      nextPage = newPageNumber;
    };
    render(
      <Pagination
        total={totalJobs}
        currentPage={currentPage}
        onPageChange={(page) => onPageChange(page)}
      />
    );
    expect(screen.getByText('Previous')).toBeInTheDocument();
    expect(screen.getByText('Next')).toBeInTheDocument();
    expect(screen.getByText('1')).toBeInTheDocument();
    expect(screen.getByText('2')).toBeInTheDocument();
    fireEvent.click(screen.getByText('Next'));
    await waitFor(() => {
      expect(nextPage).toBe(2);
    });
  });

  test('navigate next then previous page', async () => {
    const totalJobs = 30;
    const currentPage = 2;
    let newlyChangedPageNumber = 2;
    const onPageChange = (newPageNumber) => {
      newlyChangedPageNumber = newPageNumber;
    };
    render(
      <Pagination
        total={totalJobs}
        currentPage={currentPage}
        onPageChange={(page) => onPageChange(page)}
      />
    );
    expect(screen.getByText('Previous')).toBeInTheDocument();
    expect(screen.getByText('Next')).toBeInTheDocument();
    expect(screen.getByText('1')).toBeInTheDocument();
    expect(screen.getByText('2')).toBeInTheDocument();
    fireEvent.click(screen.getByText('Next'));
    await waitFor(() => {
      expect(newlyChangedPageNumber).toBe(2);
    });
    fireEvent.click(screen.getByText('Previous'));
    await waitFor(() => {
      expect(newlyChangedPageNumber).toBe(1);
    });
  });

  test('navigate next then previous page and stop', async () => {
    const totalJobs = 30;
    const currentPage = 2;
    let newlyChangedPageNumber = 2;
    const onPageChange = (newPageNumber) => {
      newlyChangedPageNumber = newPageNumber;
    };
    render(
      <Pagination
        total={totalJobs}
        currentPage={currentPage}
        onPageChange={(page) => onPageChange(page)}
      />
    );
    expect(screen.getByText('Previous')).toBeInTheDocument();
    expect(screen.getByText('Next')).toBeInTheDocument();
    expect(screen.getByText('1')).toBeInTheDocument();
    expect(screen.getByText('2')).toBeInTheDocument();
    fireEvent.click(screen.getByText('Next'));
    await waitFor(() => {
      expect(newlyChangedPageNumber).toBe(2);
    });
    fireEvent.click(screen.getByText('Previous'));
    await waitFor(() => {
      expect(newlyChangedPageNumber).toBe(1);
    });
    fireEvent.click(screen.getByText('Previous'));
    await waitFor(() => {
      expect(newlyChangedPageNumber).toBe(1);
    });
  });

  test('navigate next and stop', async () => {
    const totalJobs = 30;
    const currentPage = 1;
    let nextPage = 1;
    const onPageChange = (newPageNumber) => {
      nextPage = newPageNumber;
    };
    render(
      <Pagination
        total={totalJobs}
        currentPage={currentPage}
        onPageChange={(page) => onPageChange(page)}
      />
    );
    expect(screen.getByText('Previous')).toBeInTheDocument();
    expect(screen.getByText('Next')).toBeInTheDocument();
    expect(screen.getByText('1')).toBeInTheDocument();
    expect(screen.getByText('2')).toBeInTheDocument();
    fireEvent.click(screen.getByText('Next'));
    await waitFor(() => {
      expect(nextPage).toBe(2);
    });
    fireEvent.click(screen.getByText('Next'));
    await waitFor(() => {
      expect(nextPage).toBe(2);
    });
    screen.debug();
  });
});
