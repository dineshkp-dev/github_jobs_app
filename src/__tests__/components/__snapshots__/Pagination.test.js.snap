// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Pagination snapshot test renders 1`] = `
<div>
  <div
    class="pagination-container"
  >
    <div
      class="button-disabled"
    >
      <div
        class="button-text"
      >
        Previous
      </div>
    </div>
    <div
      class="pagination-class-indicator"
    >
      <div
        class="pagination-page active-page"
      >
        1
      </div>
    </div>
    <div
      class="button-disabled"
    >
      <div
        class="button-text"
      >
        Next
      </div>
    </div>
  </div>
</div>
`;
